Docker/validate.py

```diff
@@ -0,0 +1,158 @@
+                    member.name = os.path.basename(member.name)
+                    tar_ref.extract(member)
+                    names.append(member.name)
+    return names
+
+
+def _validate_scRNA(ds_files, pred_files):
+    """validate on scRNA-seq data"""
+    invalid_reasons = []
+    # validate each prediction file
+    for index, pred_f in enumerate(pred_files):
+        # read ds file
+        ds_df = pd.read_csv(ds_files[index], index_col=0)
+        # read corresponding pred file
+        pred_df = pd.read_csv(pred_f, index_col=0)
+        # check if all genes exist in predictions
+        cp1 = set(list(pred_df.index)).issubset(list(ds_df.index))
+        # check if all cells exist in predictions
+        cp2 = set(list(pred_df.columns.values)).issubset(
+            list(ds_df.columns.values))
+        if not (cp1 and cp2):
+            invalid_reasons.append(
+                pred_f + ": Do not contain all genes or cells")
+        else:
+            # check if all values are numeric
+            cp3 = pred_df.apply(lambda s: pd.to_numeric(
+                s, errors='coerce').isnull().any())
+            if cp3.any():
+                invalid_reasons.append(
+                    pred_f + ": Not all values are numeric")
+            # check if all values are >= 0
+            elif (pred_df < 0).any().any():
+                invalid_reasons.append(
+                    pred_f + ": Negative value is not allowed")
+    return invalid_reasons
+
+
+def main():
+    """Main function."""
+    args = get_args()
+
+    invalid_reasons = []
+
+    # set variables to find files
+    ds_props = args.proportion
+    conditions = args.condition
+    file_prefix = args.file_prefix
+
+    # check if all required downsampled data exists
+    true_ds_fs = [file_prefix + "_" + c + "_" + p + ".csv"
+                  for p in ds_props for c in conditions]
+    # downsampled files should be copied to working dir
+    diff = list(set(true_ds_fs) - set(os.listdir(".")))
+    if diff:
+        invalid_reasons.append("File not found : " + "', '".join(diff))
+
+    # validate prediction file
+    if args.submission_file is None:
+        invalid_reasons.append(
+            'Expected FileEntity type but found ' + args.entity_type
+        )
+    else:
+        # decompress submission file
+        pred_fs = _decompress_file(args.submission_file)
+        true_pred_fs = [file_prefix + "_" + c + "_" + p + "_imputed.csv"
+                        for p in ds_props for c in conditions]
+        # check if all required data exists
+        diff = list(set(true_pred_fs) - set(pred_fs))
+        if diff:
+            invalid_reasons.append("File not found : " + "', '".join(diff))
+
+    if not invalid_reasons:
+        if args.question == "1A":
+            # validate predicted data
+            scRNA_res = _validate_scRNA(true_ds_fs, true_pred_fs)
+            invalid_reasons.extend(scRNA_res)
+        elif args.question == "1B":
```

Comment:
I forget, will the challenge have more questions other than these 1A and 1B?  If not, it will be better practice to use `else` here.

Reply:
Oh, that's fine then!  I just wasn't sure if more questions were to be added.  Sometimes, `if`/`elif` is appropriate.  But in this case, the question number has to be something, hence why I recommended the `if`/`elif`/`else` approach.